### PR TITLE
[SPARK-14296][SQL] whole stage codegen support for Dataset.map

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

This PR adds a new operator `MapElements` for `Dataset.map`, it's a 1-1 mapping and is easier to adapt to whole stage codegen framework. 
## How was this patch tested?

new test in `WholeStageCodegenSuite`

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @davies @marmbrus

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala [hunk 1]
```diff
@@ -91,6 +91,32 @@ case class MapPartitions(
...
+}
+
+/**
+ * A relation produced by applying `func` to each element of the `child`.
+ *
+ * @param deserializer used to extract the input to `func` from an input row.
+ * @param serializer use to serialize the output of `func`.
+ */
+case class MapElements(
+    func: Any => Any,
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala]: this should just be AnyRef? then you can use codegen to actually call the java version without wrapping the scala one.

**Comment by cloud-fan** [comment]
generated code for
```
case class Data(i: Int, j: Long, k: String)
sqlContext.range(10).map(i => Data(i.toInt, i, i.toString))
```
looks like:
```
/* 092 */       /*** CONSUME: MapElements <function1>, newInstance(class java.lang.Long), [input[0, object].i AS i#6,input[0, object].j AS j#7L,staticinvoke(c... */
/* 093 */
/* 094 */       /* <function1>.apply */
/* 095 */       /* <function1> */
/* 096 */       /* expression: <function1> */
/* 097 */       Object mapelements_obj = ((Expression) references[1]).eval(null);
/* 098 */       scala.Function1 mapelements_value1 = (scala.Function1) mapelements_obj;
/* 099 */       /* newInstance(class java.lang.Long) */
/* 100 */       boolean mapelements_isNull2 = true;
/* 101 */       java.lang.Long mapelements_value2 = null;
/* 102 */       if (!(false)) {
/* 103 */         mapelements_value2 = new java.lang.Long(range_value);
/* 104 */         mapelements_isNull2 = false;
/* 105 */       }
/* 106 */
/* 107 */       boolean mapelements_isNull = false;
/* 108 */       org.apache.spark.sql.execution.Data mapelements_value =
/* 109 */       mapelements_isNull ?
/* 110 */       null : (org.apache.spark.sql.execution.Data) mapelements_value1.apply(mapelements_value2);
/* 111 */       mapelements_isNull = mapelements_value == null;
/* 112 */
/* 113 */       /*** CONSUME: WholeStageCodegen */
/* 114 */
/* 115 */       /* lambdavariable(mapelements_value, mapelements_isNull, ObjectType(class org.apache.spark.sql.execution.Data)).i */
/* 116 */       boolean mapelements_isNull4 = mapelements_isNull;
/* 117 */       int mapelements_value4 =
/* 118 */       mapelements_isNull4 ?
/* 119 */       -1 : (int) mapelements_value.i();
/* 120 */       /* lambdavariable(mapelements_value, mapelements_isNull, ObjectType(class org.apache.spark.sql.execution.Data)).j */
/* 121 */       boolean mapelements_isNull5 = mapelements_isNull;
/* 122 */       long mapelements_value5 =
/* 123 */       mapelements_isNull5 ?
/* 124 */       -1L : (long) mapelements_value.j();
/* 125 */       /* staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, lambdavariable(mapelements_value, mapelemen... */
/* 126 */       /* lambdavariable(mapelements_value, mapelements_isNull, ObjectType(class org.apache.spark.sql.execution.Data)).k */
/* 127 */       boolean mapelements_isNull7 = mapelements_isNull;
/* 128 */       java.lang.String mapelements_value7 =
/* 129 */       mapelements_isNull7 ?
/* 130 */       null : (java.lang.String) mapelements_value.k();
/* 131 */       mapelements_isNull7 = mapelements_value7 == null;
/* 132 */
/* 133 */       boolean mapelements_isNull6 = !!(mapelements_isNull7);
/* 134 */       UTF8String mapelements_value6 = null;
/* 135 */
/* 136 */       if (!(mapelements_isNull7)) {
/* 137 */         mapelements_value6 = org.apache.spark.unsafe.types.UTF8String.fromString(mapelements_value7);
/* 138 */         mapelements_isNull6 = mapelements_value6 == null;
/* 139 */       }
/* 140 */       mapelements_holder.reset();
/* 141 */
/* 142 */       mapelements_rowWriter.zeroOutNullBytes();
/* 143 */
/* 144 */       if (mapelements_isNull4) {
/* 145 */         mapelements_rowWriter.setNullAt(0);
/* 146 */       } else {
/* 147 */         mapelements_rowWriter.write(0, mapelements_value4);
/* 148 */       }
/* 149 */
/* 150 */       if (mapelements_isNull5) {
/* 151 */         mapelements_rowWriter.setNullAt(1);
/* 152 */       } else {
/* 153 */         mapelements_rowWriter.write(1, mapelements_value5);
/* 154 */       }
/* 155 */
/* 156 */       if (mapelements_isNull6) {
/* 157 */         mapelements_rowWriter.setNullAt(2);
/* 158 */       } else {
/* 159 */         mapelements_rowWriter.write(2, mapelements_value6);
/* 160 */       }
/* 161 */       mapelements_result.setTotalSize(mapelements_holder.totalSize());
/* 162 */       append(mapelements_result);
/* 163 */
/* 164 */       if (shouldStop()) return;
```

File: core/src/main/java/org/apache/spark/api/java/function/MapFunction.java [hunk 2]
```diff
@@ -23,5 +23,5 @@
  * Base interface for a map function used in Dataset's map function.
  */
 public interface MapFunction<T, U> extends Serializable {
-  U call(T value) throws Exception;
```
Comment (cloud-fan) [inline core/src/main/java/org/apache/spark/api/java/function/MapFunction.java]: These function APIs will be implemented by users and called by us, why do we declare it to throw exceptions? It brings a lot of trouble when using it in codegen, but seems no benefit?  cc @rxin

Comment (rxin) [inline core/src/main/java/org/apache/spark/api/java/function/MapFunction.java]: If they don't, then users must catch all exceptions and can never throw checked exceptions anymore. We shouldn't change this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala [hunk 3]
```diff
@@ -68,6 +71,64 @@ case class MapPartitions(
 }
 
 /**
+ * Applies the given function to each input row and encodes the result.
+ */
+case class MapElements(
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: it'd be great to document why this is not just a project

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala [hunk 4]
```diff
@@ -68,6 +71,69 @@ case class MapPartitions(
...
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    val callFunc: Any => Any = func match {
+      case m: MapFunction[_, _] => i => m.asInstanceOf[MapFunction[Any, Any]].call(i)
+      case _ => func.asInstanceOf[Any => Any]
+    }
+    child.execute().mapPartitionsInternal { iter =>
+      val getObject = generateToObject(deserializer, child.output)
+      val outputObject = generateToRow(serializer)
+      iter.map(getObject).map(callFunc).map(outputObject)
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: It's better to combine them into a single `map` to avoid unnecessary iterators.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala [hunk 5]
```diff
@@ -68,6 +71,69 @@ case class MapPartitions(
 }
 
 /**
+ * Applies the given function to each input row and encodes the result.
+ *
+ * TODO: Each serializer expression needs the result object which is returned by the given function,
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: I would soften this.  Its not clear that removing this small amount of code duplication is worth the confusion that would result in showing the un-common-subexpression-eliminated version of a project in the explain.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 6]
```diff
@@ -1882,7 +1882,12 @@ class Dataset[T] private[sql](
    * @since 1.6.0
    */
   @Experimental
-  def map[U : Encoder](func: T => U): Dataset[U] = mapPartitions(_.map(func))
+  def map[U : Encoder](func: T => U): Dataset[U] = {
+    new Dataset[U](
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: `withTypedPlan`?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 7]
```diff
@@ -1893,8 +1898,13 @@ class Dataset[T] private[sql](
    * @since 1.6.0
    */
   @Experimental
-  def map[U](func: MapFunction[T, U], encoder: Encoder[U]): Dataset[U] =
-    map(t => func.call(t))(encoder)
+  def map[U](func: MapFunction[T, U], encoder: Encoder[U]): Dataset[U] = {
+    implicit val uEnc = encoder
+    new Dataset[U](
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: `withTypedPlan`?

**Comment by davies** [comment]
LGTM. This implementation does not require common subexpression elimination, right?

**Comment by cloud-fan** [comment]
@davies , yea, no need of subexpression elimination for this PR.

**Comment by cloud-fan** [comment]
benchmark added, the result is included in the benchmark code. I also ran this benchmark against master branch, the result is:
```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_60-b27 on Mac OS X 10.11.4
Intel(R) Core(TM) i7-4960HQ CPU @ 2.60GHz
Dataset.map:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
-------------------------------------------------------------------------------------------
scala function                           2471 / 2531          4.0         247.1       1.0X
java function                            2416 / 2478          4.1         241.6       1.0X
```

So with whole stage codegen, we can get about 2.5 times speed up!

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+  def main(args: Array[String]): Unit = {
+    val sparkContext = new SparkContext("local[*]", "benchmark")
+    val sqlContext = new SQLContext(sparkContext)
+
+    import sqlContext.implicits._
+
+    val numRows = 10000000
+    val ds = sqlContext.range(numRows).map(l => Data(l.toInt, l.toString))
+    ds.cache()
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala]: Why cache here? uncached range() should be faster than cached.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+      res.queryExecution.toRdd.count()
+    }
+
+    /*
+    Java HotSpot(TM) 64-Bit Server VM 1.8.0_60-b27 on Mac OS X 10.11.4
+    Intel(R) Core(TM) i7-4960HQ CPU @ 2.60GHz
+    Dataset.map:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
+    -------------------------------------------------------------------------------------------
+    scala function                           1029 / 1080          9.7         102.9       1.0X
+    java function                             965 /  999         10.4          96.5       1.1X
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala]: This is very slow, range/filter/aggregate just take a few nano seconds.

Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala]: Could you have one more test case to show chained function ?

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    }
+    val enc = implicitly[Encoder[Data]]
+    benchmark.addCase("java function") { iter =>
+      var res = ds
+      var i = 0
+      while (i < 10) {
+        res = res.map(javaFunc, enc)
+        i += 1
+      }
+      res.queryExecution.toRdd.count()
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala]: This is doing the expensive conversion to external rows.  Could you try to run the existing benchmark which avoids this and also compares against RDDs?

Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala]: Okay... its harder than I thought to run it on master :(

At least base the benchmark on these tests: https://github.com/databricks/spark-sql-perf/blob/master/src/main/scala/com/databricks/spark/sql/perf/DatasetPerformance.scala

(back to back maps, compare with RDDs)

Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala]: +1 for comparing with RDDs. Let's also compare it with DF version?

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    /*
+    Java HotSpot(TM) 64-Bit Server VM 1.8.0_60-b27 on Mac OS X 10.11.4
+    Intel(R) Core(TM) i7-4960HQ CPU @ 2.60GHz
+    back-to-back map:                   Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
+    -------------------------------------------------------------------------------------------
+    Dataset                                   902 /  995         11.1          90.2       1.0X
+    DataFrame                                 132 /  167         75.5          13.2       6.8X
+    RDD                                       216 /  237         46.3          21.6       4.2X
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala]: `Dataset` is slower than `RDD` for this case, as it has some extra deserialization and serialization.

Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala]: cc @marmbrus not part of this pr, but can we reuse the same object in dataset?

Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala]: Yeah, we did this in the initial prototype of encoders.

Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala]: maybe we can have a sql config flag that determines whether the objects can be reused

Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala]: It would be really cool if you could just do this and the we would reuse:

```scala
case class Data(var l: Long, var s: String)
```

Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala]: hmmm that's mixing semantics --- val or var is only an indication of mutability, not whether the underlying object is the same across records ...

Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala]: And the reason that you would make a case class mutable other than to allow reuse is?  You could have another conf if you really wanted to allow them to prevent reuse / for Java where it is less clear.

However they have to make them vars anyway and I think the intention is pretty clear if they do so.  Requiring them to lookup some conf seems onerous, so I'd probably make reuse the default when possible.

Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetBenchmark.scala]: "And the reason that you would make a case class mutable other than to allow reuse is"

because you might want to mutate state? this shouldn't work only for case classes -- there are other things we should reuse, e.g. tuples, java beans. I wouldn't want to overload an existing language feature that is not about object reuse.

I do agree it'd be better to reuse by default for performance improvements, and users can explicitly turn it off through a flag.

**Comment by cloud-fan** [comment]
merging to master!
